iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,955 @@
+
+  refcountcache_last_sync_time,   // seconds since epoch of last successful sync
+  refcountcache_last_total_items, // number of items sync last time
+  refcountcache_last_total_size,  // total size at last sync
+
+  RefCountCache_Stat_Count
+};
+
+struct RefCountCacheItemMeta {
+  uint64_t key;
+  unsigned int size;
+  ink_time_t expiry_time; // expire time as seconds since epoch
+  RefCountCacheItemMeta(uint64_t key, unsigned int size, int expire_time = -1) : key(key), size(size), expiry_time(expire_time) {}
+};
+
+// Layer of indirection for the hashmap-- since it needs lots of things inside of it
+// We'll also use this as the item header, for persisting objects to disk
+class RefCountCacheHashEntry
+{
+public:
+  Ptr<RefCountObj> item;
+  LINK(RefCountCacheHashEntry, item_link);
+
+  PriorityQueueEntry<RefCountCacheHashEntry *> *expiry_entry;
+
+  RefCountCacheItemMeta meta;
+  void
+  set(RefCountObj *i, uint64_t key, unsigned int size, int expire_time)
+  {
+    this->item = make_ptr(i);
+    this->meta = RefCountCacheItemMeta(key, size, expire_time);
+  };
+  // Need a no-argument constructor to use the classAllocator
+  RefCountCacheHashEntry() : expiry_entry(NULL), meta(0, 0){};
+
+  // make these values comparable -- so we can sort them
+  const bool
+  operator<(const RefCountCacheHashEntry &v2)
+  {
+    return this->meta.expiry_time < v2.meta.expiry_time;
+  };
+};
+// Since the hashing values are all fixed size, we can simply use a classAllocator to avoid mallocs
+extern ClassAllocator<RefCountCacheHashEntry> refCountCacheHashingValueAllocator;
+extern ClassAllocator<PriorityQueueEntry<RefCountCacheHashEntry *>> expiryQueueEntry;
+
+struct RefCountCacheHashing {
+  typedef uint64_t ID;
+  typedef uint64_t const Key;
+  typedef RefCountCacheHashEntry Value;
+  typedef DList(RefCountCacheHashEntry, item_link) ListHead;
+
+  static ID
+  hash(Key key)
+  {
+    return key;
+  }
+  static Key
+  key(Value const *value)
+  {
+    return value->meta.key;
+  }
+  static bool
+  equal(Key lhs, Key rhs)
+  {
+    return lhs == rhs;
+  }
+};
+
+// The RefCountCachePartition is simply a map of key -> Ptr<RefCountCacheItem<YourClass>>
+template <class C> class RefCountCachePartition
+{
+public:
+  RefCountCachePartition(uint64_t max_size, unsigned int max_items, RecRawStatBlock *rsb = NULL);
+  RefCountCachePartition(const RefCountCachePartition &);
+  Ptr<C> get(uint64_t key);
+  void put(uint64_t key, C *item, int size = 0, int expire_time = 0);
+  void erase(uint64_t key, int expiry_time = -1);
+
+  void clear();
+  bool is_full();
+  bool reserve(unsigned int);
+
+  const size_t num_items();
+  void copy(std::vector<RefCountCacheHashEntry> *items);
+  // implement an assignment operator, to appease the vector constructor in older versions of gcc
+  RefCountCachePartition &
+  operator=(const RefCountCachePartition &other)
+  {
+    if (this == &other) {
+      return *this;
+    }
+    *this = RefCountCachePartition(other);
```

Comment:
TSHashTable is guaranteed to be unsafe to copy construct. There are various reasons but the most important is that the elements can be in only one TSHashTable at a time due to the embedded links.